File: packages/contracts/src/sse.ts
import * as z from "zod";
import { Id, Instant } from "./primitives";
import { ResolutionClass } from "./concepts/resolution-class";
import { RecordRef } from "./concepts/question";

// The answer stream (flows 6, 7, 8): Server-Sent Events from `POST /v1/events` and
// `POST /v1/events/:id/initiative`. On the wire each event is `event: <name>` plus `data: <JSON payload>`;
// the payload does not repeat the name.
// Order: meta, (provisional), delta+, citation*, class, (count), done. Or `choose_initiative` alone.
// `error` replaces the tail only when even persistence failed.

export const InitiativeOption = z.object({ initiativeId: Id, name: z.string().min(1) });
export type InitiativeOption = z.infer<typeof InitiativeOption>;

export const MetaPayload = z.object({
  eventId: Id,
  initiativeId: Id,
  initiativeName: z.string(),
  /** How sure the service is that the event belongs to this initiative, 0 to 1. */
  initiativeConfidence: z.number().min(0).max(1),
  /** Friction picked the initiative and is only fairly sure: the chip offers `alternatives`. */
  switchable: z.boolean(),
  /** The employee's other live initiatives, by full name, when `switchable`. Empty otherwise. */
  alternatives: z.array(InitiativeOption),
  /** Name of the initiative the previous answer was for, after a switch. */
  switchedFrom: z.string().nullable(),
  /** Set when the event is a follow-up in a thread. */
  inReplyTo: RecordRef.nullable(),
});
/** Friction could not tell which initiative the flag is about. The stream ends after this, with no `done`. */
export const ChooseInitiativePayload = z.object({ options: z.array(InitiativeOption) });
export const ProvisionalPayload = z.object({ message: z.string().min(1) });
export const DeltaPayload = z.object({ text: z.string() });
export const CitationSourceKind = z.enum(["document", "qa"]);
export const CitationPayload = z
  .object({
    /** 1-based number shown on the chip. */
    index: z.int().positive(),
    /** The exact cited sentences. */
    quote: z.string().min(1),
    documentTitle: z.string().min(1),
    locator: z.string().nullable(),
    /** Answer text length at the end of the sentence group this citation supports: where the chip goes. */
    charOffset: z.int().nonnegative(),
    sourceKind: CitationSourceKind,
    /** "4 Purchase orders > 4.2 Approvals". Null for Q&A. */
    headingPath: z.string().nullable(),
    /** The whole passage (or the owner's answer), for "Open at passage". */
    passageText: z.string().min(1),
    /** The cited document is marked for review. */
    suspect: z.boolean(),
    qaApprovedByUserId: z.string().nullable(),
    qaApprovedAt: Instant.nullable(),
  })
  .and(
    z.xor([
      z.object({ passageId: Id, qaEntryId: z.null() }),
      z.object({ passageId: z.null(), qaEntryId: Id }),
    ]),
  );
export type CitationPayload = z.infer<typeof CitationPayload>;
export const ClassPayload = z.object({
  resolutionClass: ResolutionClass,
  /** The event went to the initiative's owners (Unanswerable, Still stuck). */
  routed: z.boolean(),
});
export const CountPayload = z.object({ othersCount: z.int().nonnegative() });
export const DonePayload = z.object({ answerId: Id });
export const ErrorPayload = z.object({ code: z.string().min(1), message: z.string().min(1) });

export const SSE_PAYLOADS = {
  meta: MetaPayload,
  choose_initiative: ChooseInitiativePayload,
  provisional: ProvisionalPayload,
  delta: DeltaPayload,
  citation: CitationPayload,
  class: ClassPayload,
  count: CountPayload,
  done: DonePayload,
  error: ErrorPayload,
} as const;

export type SSEEventName = keyof typeof SSE_PAYLOADS;
export const SSE_EVENT_NAMES = Object.keys(SSE_PAYLOADS) as SSEEventName[];

export type SSEEvent = {
  [K in SSEEventName]: { event: K; data: z.infer<(typeof SSE_PAYLOADS)[K]> };
}[SSEEventName];

/** Parses one wire event (`event` name plus JSON `data` text) into a typed event. */
export function parseSSEEvent(event: string, data: string): SSEEvent {
  if (!(event in SSE_PAYLOADS)) throw new Error(`Unknown answer stream event "${event}"`);
  const name = event as SSEEventName;
  return { event: name, data: SSE_PAYLOADS[name].parse(JSON.parse(data)) } as SSEEvent;
}

/** A scripted stream step: the event plus how long to wait before sending it. Used by the shell's stub and fixtures. */
export const ScriptStep = z.object({
  delayMs: z.int().nonnegative(),
  event: z.enum(SSE_EVENT_NAMES as [SSEEventName, ...SSEEventName[]]),
  data: z.record(z.string(), z.unknown()),
});
export type ScriptStep = z.infer<typeof ScriptStep>;

File: apps/mac/Friction/Contracts/SSE.swift
import Foundation

/// The answer stream (packages/contracts/src/sse.ts). On the wire: `event: <name>` and `data: <JSON payload>`.
nonisolated enum SSEEvent: Sendable, Equatable {
    struct Meta: Codable, Sendable, Equatable {
        let eventId: UUID
        let initiativeId: UUID
        let initiativeName: String
        let initiativeConfidence: Double
    }
    struct Provisional: Codable, Sendable, Equatable { let message: String }
    struct Delta: Codable, Sendable, Equatable { let text: String }
    struct CitationEvent: Codable, Sendable, Equatable {
        let index: Int
        let passageId: UUID?
        let qaEntryId: UUID?
        let quote: String
        let documentTitle: String
        let locator: String?
    }
    struct Class: Codable, Sendable, Equatable { let resolutionClass: ResolutionClass }
    struct Count: Codable, Sendable, Equatable { let othersCount: Int }
    struct Done: Codable, Sendable, Equatable { let answerId: UUID }
    struct Failure: Codable, Sendable, Equatable { let code: String; let message: String }

    case meta(Meta)
    case provisional(Provisional)
    case delta(Delta)
    case citation(CitationEvent)
    case resolutionClass(Class)
    case count(Count)
    case done(Done)
    case error(Failure)

    enum DecodeError: Error, Equatable { case unknownEvent(String) }

    /// Decodes one wire event. Unknown names throw, so a renamed event fails loudly rather than vanishing.
    init(name: String, data: Data) throws {
        let d = JSONDecoder.contract
        switch name {
        case "meta": self = .meta(try d.decode(Meta.self, from: data))
        case "provisional": self = .provisional(try d.decode(Provisional.self, from: data))
        case "delta": self = .delta(try d.decode(Delta.self, from: data))
        case "citation": self = .citation(try d.decode(CitationEvent.self, from: data))
        case "class": self = .resolutionClass(try d.decode(Class.self, from: data))
        case "count": self = .count(try d.decode(Count.self, from: data))
        case "done": self = .done(try d.decode(Done.self, from: data))
        case "error": self = .error(try d.decode(Failure.self, from: data))
        default: throw DecodeError.unknownEvent(name)
        }
    }
}

/// One step of a scripted stub stream (fixtures `stub-*.json`).
nonisolated struct ScriptStep: Decodable, Sendable {
    let delayMs: Int
    let event: String
    let data: JSONValue
}

/// Minimal JSON value so script payloads can be re-encoded and fed through `SSEEvent(name:data:)`.
nonisolated enum JSONValue: Codable, Sendable, Equatable {
    case string(String), number(Double), bool(Bool), object([String: JSONValue]), array([JSONValue]), null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null }
        else if let b = try? c.decode(Bool.self) { self = .bool(b) }
        else if let n = try? c.decode(Double.self) { self = .number(n) }
        else if let s = try? c.decode(String.self) { self = .string(s) }
        else if let a = try? c.decode([JSONValue].self) { self = .array(a) }
        else { self = .object(try c.decode([String: JSONValue].self)) }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let n): if n.rounded() == n, abs(n) < 1e15 { try c.encode(Int(n)) } else { try c.encode(n) }
        case .bool(let b): try c.encode(b)
        case .object(let o): try c.encode(o)
        case .array(let a): try c.encode(a)
        case .null: try c.encodeNil()
        }
    }
}
